kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/ResourcesUtil.java

```diff
@@ -288,32 +286,39 @@ static List<Condition> maybeAddOrUpdateCondition(List<Condition> conditions, Con
                 .thenComparing(Condition::getMessage)
                 .thenComparing(Condition::getLastTransitionTime);
 
-        TreeMap<Condition.Type, List<Condition>> byType = conditions.stream().collect(Collectors.groupingBy(
-                Condition::getType,
-                TreeMap::new, // order based on Type
-                Collectors.toList()));
-
-        var map = byType.entrySet().stream().map(entry -> {
-            // minimum must exist because values of groupingBy result are always non-empty
-            Condition minimumCondition = entry.getValue().stream().min(conditionComparator).orElseThrow();
-            var mutableList = new ArrayList<Condition>(1);
-            mutableList.add(minimumCondition);
-            return Map.entry(entry.getKey(), mutableList);
-        }).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue,
-                (l1, l2) -> {
-                    throw new IllegalStateException();
-                },
-                TreeMap::new));
-
-        ArrayList<Condition> conditionsOfType = map.computeIfAbsent(type, k -> new ArrayList<>());
-        if (conditionsOfType.isEmpty()) {
-            conditionsOfType.add(condition);
-        }
-        else if (conditionsOfType.get(0).getObservedGeneration() <= condition.getObservedGeneration()) {
-            conditionsOfType.set(0, condition);
+        for (Condition condition : conditionsToAdd) {
+            var type = Objects.requireNonNull(condition.getType());
+
+            Map<Condition.Type, List<Condition>> byType = conditions.stream().collect(Collectors.groupingBy(
+                    Condition::getType));
+
+            var map = byType.entrySet().stream().map(entry -> {
+                // minimum must exist because values of groupingBy result are always non-empty
+                Condition minimumCondition = entry.getValue().stream().min(conditionComparator).orElseThrow();
+                var mutableList = new ArrayList<Condition>(1);
```

Comment:
nit: the relationship between this mutable list and the later code that depends on it's mutability is a bit fiddly, I think we'd be better off breaking it into two methods doing immutable transforms.

```
Map<Condition.Type, List<Condition>> freshestConditions = selectFreshest(byType)
Map<Condition.Type, List<Condition>> updatedConditions = addCondition(condition, freshestConditions)
```